# NYC zipcode gathering steps:

In [1]:
#Importing pandas and numpy 
import pandas as pd
import numpy as np

#Below is a list of NYC zipcodes scrapped from various sources on the internet:

In [2]:
Bronx_zipcodes = [10453, 10457, 10460, 10458, 10467, 10468
,10451, 10452, 10456,10454, 10455, 10459, 10474, 10463, 10471
,10466, 10469, 10470, 10475, 10461, 10462,10464, 10465, 10472, 10473, 10803 ]

In [3]:
Brooklyn_zipcodes = [11212, 11213, 11216, 11233, 11238
, 11209, 11214, 11228, 11204, 11218, 11219, 11230,11234, 11236, 11239
,11223, 11224, 11229, 11235, 11201, 11205, 11215, 11217, 11231
, 11203, 11210, 11225, 11226, 11207, 11208,11211, 11222
, 11220, 11232, 11206, 11221, 11237, 11249]

In [4]:
Manhattan_zipcodes = [10026, 10027, 10030, 10037, 10039
,10001, 10011, 10018, 10019, 10020, 10036
,10029, 10035,10010, 10016, 10017, 10022
,10012, 10013, 10014,10004, 10005, 10006, 10007, 10038, 10280
,10002, 10003, 10009,10021, 10028, 10044, 10065, 10075, 10128
,10023, 10024, 10025,10031, 10032, 10033, 10034, 10040, 10282,10278]

In [5]:
Queens_zipcodes = [11001, 11351,11354, 11355, 11356, 11357, 11358, 11359, 11360, 11361, 11362, 11363, 11364
,11354, 11355, 11356, 11357, 11358, 11359, 11360
, 11365, 11366, 11367, 11405, 11412, 11423, 11432, 11433, 11434, 11435, 11436
, 11101, 11102, 11103, 11104, 11105, 11106, 11374, 11375, 11379, 11385
,11691, 11692, 11693, 11694, 11695, 11697,11004
,11005, 11411, 11413, 11422, 11426, 11427, 11428, 11429
, 11414, 11415, 11416, 11417, 11418, 11419, 11420, 11421
, 11368, 11369, 11370, 11371, 11372, 11373,11374,11375, 11377, 11378,11379,11385, 11405,11001,11439]


In [6]:
Staten_zipcodes = [10302, 10303, 10310,10306, 10307,10313,  10308, 10309, 10312,10301, 10304, 10305,10311,10314]

Below we load data from pluto, then proceed to gather all the unique zipcodes found:

In [7]:
pluto = pd.read_csv('pluto_20v1.csv')
        

C:\Users\jabakoyas\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (17,18,20,22) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [8]:
pluto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 859172 entries, 0 to 859171
Data columns (total 99 columns):
borough                 859171 non-null object
block                   859172 non-null int64
lot                     859172 non-null int64
cd                      858186 non-null float64
ct2010                  858186 non-null float64
cb2010                  858186 non-null float64
schooldist              857481 non-null float64
council                 858185 non-null float64
zipcode                 857449 non-null float64
firecomp                857464 non-null object
policeprct              857473 non-null float64
healtharea              857474 non-null float64
sanitboro               857270 non-null float64
sanitsub                857150 non-null object
address                 858772 non-null object
zonedist1               856974 non-null object
zonedist2               19787 non-null object
zonedist3               207 non-null object
zonedist4               13 non-null obje

In [9]:
#We create a zipcode_dataset by subsetting only the borough and zipcode data from pluto
zipcode_dataset = pluto[['borough', 'zipcode']]

In [10]:
#We remove rows with any missing values
zipcode_dataset = zipcode_dataset[np.logical_not(np.isnan(zipcode_dataset['zipcode']))]

In [11]:
#Next we select all the unique zipcodes
zipcode_dataset = zipcode_dataset.drop_duplicates(subset='zipcode', keep= 'first')
zipcode_dataset

borough  zipcode
0           BK  11220.0
1           QN  11356.0
2           BK  11203.0
3           BK  11234.0
8           BK  11235.0
...        ...      ...
769066      MN  10045.0
785175      QN  11451.0
803135      MN  10176.0
826286      MN  10151.0
827569      MN  10169.0

[215 rows x 2 columns]

Above we see that we have 215 unique zipcodes from Pluto

In [12]:
#Next we reset the index for better presentation
zipcode_dataset.reset_index(inplace=True, drop=True)

In [13]:
#And we change the zipcodes to integers by removing the decimal points (and also the zero after the decimal)
zipcode_dataset['zipcode'] = zipcode_dataset['zipcode'].astype(int)

In [14]:
#And here we sort the rows by the borough
zipcode_dataset = zipcode_dataset.sort_values(by=['borough'])
zipcode_dataset

borough  zipcode
0        BK    11220
56       BK    11225
28       BK    11236
90       BK    11206
31       BK    11213
..      ...      ...
126      SI    10308
12       SI    10304
53       SI    10306
68       SI    10301
54       SI    10309

[215 rows x 2 columns]

Next, we see if there are any zipcodes from the internet gathered zipcodes that are not in Pluto

In [15]:
my_zips = Bronx_zipcodes+Brooklyn_zipcodes+Manhattan_zipcodes+Staten_zipcodes+Queens_zipcodes

In [16]:
#The list of zipcodes from the internet is slightly shorter than the Pluto list.
len(my_zips)

202

In [17]:
#We convert the zipcodes into sets so that we can do set comparisons in the next line (faster for computation than loops)
pluto_zips = set(zipcode_dataset.zipcode)
my_zips = set(my_zips)

In [18]:
#We look at the set difference of pluto zips not found in the internet list
pluto_zips.difference(my_zips)

{10045,
 10055,
 10069,
 10103,
 10105,
 10106,
 10110,
 10118,
 10119,
 10120,
 10121,
 10122,
 10123,
 10151,
 10152,
 10155,
 10158,
 10165,
 10169,
 10170,
 10175,
 10176,
 10271,
 10281,
 11040,
 11096,
 11109,
 11241,
 11251,
 11430,
 11451,
 12345}

In [19]:
#And here we do the opposite to look at zipcodes found from internet not in the Pluto list
zips = my_zips.difference(pluto_zips)
zips

{10311, 10313, 10803, 11351, 11405, 11439}

In [20]:
#Here we just create a borough, zipcode pair for the above 6 zipcodes in order to organize them and add them into the Pluto zipcodes
zips = my_zips.difference(pluto_zips)
string = ['BX', 'BK', 'MN', 'SI', 'QN']
boroughs = [Bronx_zipcodes, Brooklyn_zipcodes, Manhattan_zipcodes, Staten_zipcodes, Queens_zipcodes]
lists = []
for x in zips:
    for i, y in enumerate(boroughs):
        if x in y:
            lists.append((string[i], x))
lists


[('SI', 10311),
 ('SI', 10313),
 ('QN', 11405),
 ('QN', 11439),
 ('BX', 10803),
 ('QN', 11351)]

In [21]:
#Here we convert the above short list into a dataframe
short_list = pd.DataFrame(lists, columns=['borough', 'zipcode'])
short_list

borough  zipcode
0      SI    10311
1      SI    10313
2      QN    11405
3      QN    11439
4      BX    10803
5      QN    11351

In [22]:
#And here we merge the two dataframes togther.
zipcode_dataset = pd.concat([zipcode_dataset, short_list])
zipcode_dataset = zipcode_dataset.sort_values(by=['borough'])

In [23]:
#We reset the index once more for better presentation
zipcode_dataset.reset_index(inplace=True, drop=True)
zipcode_dataset

borough  zipcode
0        BK    11220
1        BK    11233
2        BK    11212
3        BK    11215
4        BK    11211
..      ...      ...
216      SI    10305
217      SI    10310
218      SI    10302
219      SI    10304
220      SI    10314

[221 rows x 2 columns]

In [24]:
#Finally, we store the new zipcode dataset into a csv file
zipcode_dataset.to_csv(r'C:\Users\jabakoyas\Downloads\zipcodes_nyc_complete.csv')
